hypershift-operator/controllers/nodepool/nodepool_controller.go

```diff
@@ -2917,3 +2918,140 @@ func aggregateMachineMessages(msgs []string) string {
+			log.Info("not found", "request", req.String())
+			return ctrl.Result{}, nil
+		}
+		log.Error(err, "error getting secret")
+		return ctrl.Result{}, err
+	}
+
+	nodePoolName, annotated := secret.Annotations[nodePoolAnnotation]
+	if !annotated {
+		return ctrl.Result{}, nil
+	}
+	log = log.WithValues("nodePool", nodePoolName)
+
+	nodePool := &hyperv1.NodePool{}
+	if err := r.Client.Get(ctx, supportutil.ParseNamespacedName(nodePoolName), nodePool); err != nil && !apierrors.IsNotFound(err) {
+		log.Error(err, "error getting nodepool")
+		return ctrl.Result{}, err
+	} else if apierrors.IsNotFound(err) {
+		log.Info("removing secret as nodePool is missing")
+		return ctrl.Result{}, r.Client.Delete(ctx, secret)
+	}
+
+	hcluster, err := GetHostedClusterByName(ctx, r.Client, nodePool.GetNamespace(), nodePool.Spec.ClusterName)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	controlPlaneNamespace := manifests.HostedControlPlaneNamespace(hcluster.Namespace, hcluster.Name)
+	expectedCoreConfigResources := expectedCoreConfigResourcesForHostedCluster(hcluster)
+	releaseImage, err := r.getReleaseImage(ctx, hcluster, nodePool.Status.Version, nodePool.Spec.Release.Image)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	config, missingConfigs, err := r.getConfig(ctx, nodePool, expectedCoreConfigResources, controlPlaneNamespace, releaseImage, hcluster)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if missingConfigs {
+		return ctrl.Result{}, nil
+	}
+	targetVersion := releaseImage.Version()
+
+	pullSecretName, err := r.getPullSecretName(ctx, hcluster)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	globalConfig, err := globalConfigString(hcluster)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	targetPayloadConfigHash := payloadConfigHash(config, targetVersion, pullSecretName, globalConfig)
+
+	// synchronously deleting the ignition token is unsafe; we need to clean up tokens by annotating them to expire
+	synchronousCleanup := func(ctx context.Context, c client.Client, secret *corev1.Secret) error {
+		return c.Delete(ctx, secret)
+	}
+	type nodePoolSecret struct {
+		expectedName   string
+		matchingPrefix string
+		cleanup        func(context.Context, client.Client, *corev1.Secret) error
+	}
+	valid := false
+	options := []nodePoolSecret{
+		{
+			expectedName:   TokenSecret(controlPlaneNamespace, nodePool.Name, targetPayloadConfigHash).Name,
+			matchingPrefix: tokenSecretPrefix,
+			cleanup:        setExpirationTimestampOnToken,
+		},
+		{
+			expectedName:   IgnitionUserDataSecret(controlPlaneNamespace, nodePool.GetName(), targetPayloadConfigHash).Name,
+			matchingPrefix: ignitionUserDataPrefix,
+			cleanup:        synchronousCleanup,
+		},
+	}
+	cleanup := synchronousCleanup
+	var names []string
+	for _, option := range options {
+		names = append(names, option.expectedName)
+		if secret.Name == option.expectedName {
+			valid = true
+		}
+		if strings.HasPrefix(secret.Name, option.matchingPrefix) {
```

Comment:
a "token" prefix seems to generic as to be a valid criteria, may be isTokenSecret could check all the expected annotations keys in the secret?

Reply:
Chatted on Slack - we need the prefix _and_ to have the HyperShift-specific NodePool annotation, which is the best we can do unless we add an explicit marker and backfill it to existing deployments.